File: web-ahp-app/backend/comparisons/scales.py
"""
Scale implementations for pairwise comparisons.
Ported from the original Tkinter application.
"""
import math
import numpy as np


class Scale:
    """Base class for all scale types."""

    def __init__(self, gradations=9):
        """
        Initialize scale with number of gradations.

        Args:
            gradations: Number of gradations (3-9)
        """
        self.gradations = min(max(gradations, 3), 9)
        self.values = self._calculate_values()

    def _calculate_values(self):
        """Calculate scale values. Must be implemented by subclasses."""
        raise NotImplementedError

    def unify(self, gradation_index):
        """
        Unify gradation to cardinal scale (1.5 to 9.5).

        Args:
            gradation_index: Index of gradation (1 to n)

        Returns:
            Unified value in range [1.5, 9.5]
        """
        n = self.gradations
        l = 1.5  # Lower bound
        p = 9.5  # Upper bound

        # M_i^n = l + (i - 0.5) * (p - l) / n
        unified = l + (gradation_index - 0.5) * (p - l) / n
        return unified


class IntegerScale(Scale):
    """Integer scale (1-9)."""

    def _calculate_values(self):
        """Generate integer values 1 through gradations."""
        return list(range(1, self.gradations + 1))


class BalancedScale(Scale):
    """Balanced scale using weight ratio formula."""

    def _calculate_values(self):
        """
        Generate balanced scale values.
        Formula: a = w / (1 - w), where w = i / (n + 1)
        """
        n = self.gradations
        values = []

        for i in range(1, n + 1):
            w = i / (n + 1)
            if w < 1:
                a = w / (1 - w)
            else:
                a = 9.0
            values.append(a)

        return values


class PowerScale(Scale):
    """Power scale using exponential growth."""

    def _calculate_values(self):
        """
        Generate power scale values.
        Formula: a = 9^((i-1)/(n-1))
        """
        n = self.gradations
        values = []

        for i in range(1, n + 1):
            if n > 1:
                exponent = (i - 1) / (n - 1)
                a = math.pow(9, exponent)
            else:
                a = 1.0
            values.append(a)

        return values


class MaZhengScale(Scale):
    """Ma-Zheng scale."""

    def _calculate_values(self):
        """
        Generate Ma-Zheng scale values.
        Formula: a = 9 / (9 + 1 - grade)
        """
        values = []

        for grade in range(1, self.gradations + 1):
            a = 9 / (9 + 1 - grade)
            values.append(a)

        return values


class DoneganScale(Scale):
    """Donegan scale using hyperbolic tangent."""

    def _calculate_values(self):
        """
        Generate Donegan scale values.
        Formula: a = exp(atanh((grade-1)/14 * sqrt(3)))
        """
        values = []

        for grade in range(1, self.gradations + 1):
            try:
                # atanh argument must be in (-1, 1)
                arg = (grade - 1) / 14 * math.sqrt(3)
                arg = min(max(arg, -0.999), 0.999)  # Clamp to valid range

                a = math.exp(math.atanh(arg))
                values.append(a)
            except (ValueError, OverflowError):
                values.append(1.0)

        return values


def get_scale(scale_type, gradations=9):
    """
    Factory function to get appropriate scale instance.

    Args:
        scale_type: Integer 1-5 representing scale type
        gradations: Number of gradations (3-9)

    Returns:
        Scale instance
    """
    scale_classes = {
        1: IntegerScale,
        2: BalancedScale,
        3: PowerScale,
        4: MaZhengScale,
        5: DoneganScale,
    }

    scale_class = scale_classes.get(scale_type, IntegerScale)
    return scale_class(gradations)


def integer_by_scale(grade, scale_type):
    """
    Transform a grade (1-9) using the specified scale type.

    Args:
        grade: Grade value (1-9)
        scale_type: Scale type (1-5)

    Returns:
        Transformed value
    """
    grade = max(1, min(9, grade))  # Clamp to 1-9

    if scale_type == 1:  # Integer
        return float(grade)

    elif scale_type == 2:  # Balanced
        w = 0.5 + (grade - 1) * 0.05
        if w < 1:
            return w / (1 - w)
        return 9.0

    elif scale_type == 3:  # Power
        return math.pow(9, (grade - 1) / 8)

    elif scale_type == 4:  # Ma-Zheng
        return 9 / (9 + 1 - grade)

    elif scale_type == 5:  # Donegan
        try:
            arg = (grade - 1) / 14 * math.sqrt(3)
            arg = min(max(arg, -0.999), 0.999)
            return math.exp(math.atanh(arg))
        except (ValueError, OverflowError):
            return 1.0

    return float(grade)


def get_progressive_labels(gradations):
    """
    Get labels for progressive refinement scales.

    Args:
        gradations: Number of gradations (3-9)

    Returns:
        List of label strings
    """
    # Base labels for 3 gradations
    base_labels = ["Weak", "Strong", "Extreme"]

    # Progressive expansion
    labels_map = {
        3: ["Weak", "Strong", "Extreme"],
        4: ["Weak", "Medium", "Strong", "Extreme"],
        5: ["Weak", "Medium", "Above Medium", "Strong", "Extreme"],
        6: ["Weak", "Medium", "Above Medium", "Strong", "Very Strong", "Extreme"],
        7: ["Weak", "Medium", "Above Medium", "Strong", "Very Strong", "Very Very Strong", "Extreme"],
        8: ["Weak", "Medium", "Above Medium", "Strong", "Very Strong", "Very Very Strong", "Extremely Strong", "Absolute"],
        9: ["Weak", "Medium", "Above Medium", "Strong", "Very Strong", "Very Very Strong", "Extremely Strong", "Nearly Absolute", "Absolute"],
    }

    return labels_map.get(gradations, base_labels)

File: web-ahp-app/frontend/src/utils/scales.js
/**
 * Scale calculation utilities
 * Ported from the original Python application
 */

/**
 * Unify gradation index to cardinal scale (1.5 to 9.5)
 *
 * @param {number} gradationIndex - Index of gradation (1 to n)
 * @param {number} totalGradations - Total number of gradations
 * @returns {number} Unified value in range [1.5, 9.5]
 */
export function unifyGradation(gradationIndex, totalGradations) {
  const l = 1.5 // Lower bound
  const p = 9.5 // Upper bound

  // M_i^n = l + (i - 0.5) * (p - l) / n
  const unified = l + (gradationIndex - 0.5) * (p - l) / totalGradations

  return unified
}

/**
 * Transform a grade using the specified scale type
 *
 * @param {number} grade - Grade value (1-9)
 * @param {number} scaleType - Scale type (1-5)
 * @returns {number} Transformed value
 */
export function integerByScale(grade, scaleType) {
  // Clamp grade to 1-9
  grade = Math.max(1, Math.min(9, grade))

  switch (scaleType) {
    case 1: // Integer
      return grade

    case 2: // Balanced
      {
        const w = 0.5 + (grade - 1) * 0.05
        if (w < 1) {
          return w / (1 - w)
        }
        return 9.0
      }

    case 3: // Power
      return Math.pow(9, (grade - 1) / 8)

    case 4: // Ma-Zheng
      return 9 / (9 + 1 - grade)

    case 5: // Donegan
      {
        try {
          let arg = (grade - 1) / 14 * Math.sqrt(3)
          arg = Math.max(-0.999, Math.min(0.999, arg)) // Clamp
          return Math.exp(Math.atanh(arg))
        } catch (e) {
          return 1.0
        }
      }

    default:
      return grade
  }
}

/**
 * Get progressive labels for gradations
 *
 * @param {number} gradations - Number of gradations (3-9)
 * @returns {Array<string>} Array of label strings
 */
export function getProgressiveLabels(gradations) {
  const labelsMap = {
    3: ['Weak', 'Strong', 'Extreme'],
    4: ['Weak', 'Medium', 'Strong', 'Extreme'],
    5: ['Weak', 'Medium', 'Above Medium', 'Strong', 'Extreme'],
    6: ['Weak', 'Medium', 'Above Medium', 'Strong', 'Very Strong', 'Extreme'],
    7: ['Weak', 'Medium', 'Above Medium', 'Strong', 'Very Strong', 'Very Very Strong', 'Extreme'],
    8: ['Weak', 'Medium', 'Above Medium', 'Strong', 'Very Strong', 'Very Very Strong', 'Extremely Strong', 'Absolute'],
    9: ['Weak', 'Medium', 'Above Medium', 'Strong', 'Very Strong', 'Very Very Strong', 'Extremely Strong', 'Nearly Absolute', 'Absolute'],
  }

  return labelsMap[gradations] || labelsMap[3]
}

/**
 * Get the initial scale string based on number of gradations
 * Implements the GRADUAL_SCALE mapping from original implementation
 *
 * @param {number} gradations - Number of gradations (2-8)
 * @returns {string} Scale string representing grades
 */
export function getGradualScale(gradations) {
  const scaleMap = {
    2: '25',
    3: '259',        // Initial: Weak, Strong, Extreme
    4: '3579',
    5: '23579',
    6: '234579',
    7: '2345679',
    8: '23456789'
  }

  return scaleMap[gradations] || '259'
}

/**
 * Calculate proportional button widths for scale grades
 * Matches the original Tkinter implementation
 *
 * @param {string} scaleStr - Scale string (e.g., "259", "23579")
 * @param {number} scaleType - Scale type (1-5)
 * @returns {Array<number>} Array of width percentages
 */
export function calculateButtonWidths(scaleStr, scaleType) {
  const grades = scaleStr.split('').map(c => parseInt(c))
  const li = grades.length

  // For Integer scale: equal widths
  if (scaleType === 1) {
    const equalWidth = 100 / li
    return grades.map(() => equalWidth)
  }

  // For other scales: proportional widths based on transformed values
  // Calculate sum of all transformed values
  let sum_w = 0
  for (let i = 1; i <= li; i++) {
    const unified = 1.5 + (i - 0.5) * (9.5 - 1.5) / li
    sum_w += integerByScale(unified, scaleType)
  }

  // Calculate width for each button (in reverse order from Tkinter)
  const widths = []
  for (let i = 0; i < li; i++) {
    const unified = 1.5 + (li - i - 0.5) * (9.5 - 1.5) / li
    const transformedValue = integerByScale(unified, scaleType)
    const widthPercent = (transformedValue / sum_w) * 100
    widths.push(widthPercent)
  }

  // Reverse to match button order
  return widths.reverse()
}

/**
 * Math.atanh polyfill for older browsers
 */
if (!Math.atanh) {
  Math.atanh = function(x) {
    return 0.5 * Math.log((1 + x) / (1 - x))
  }
}
